chore: remove oldSectionVars in FHE

goens opened the pull request:
introduces sections more carefully to manage typeclasses in FHE, and gets rid of oldSectionVars this way

goens committed:
chore: remove oldSectionVars in FHE, introduce sections more carefully to manage typeclasses

goens committed:
forgot Statements file

tobiasgrosser commented:
This looks very nice. @bollu, a review would be much appreciated.

bollu commented on SSA/Projects/FullyHomomorphicEncryption/Basic.lean:
```diff
@@ -607,15 +623,15 @@ theorem R.coeff_fromPoly {q n : Nat} [Fact (q > 1)] (p : (ZMod q)[X]) :
     (R.fromTensor (q := q) (n := n) tensor).coeff i = (tensor.getD i 0) := by
   rw [fromTensor_eq_fromTensorFinsupp_fromPoly]
-  have hfromTensorFinsuppDegree := fromTensorFinsupp_degree q tensor
+  have hfromTensorFinsuppDegree := fromTensorFinsupp_degree (q:=q) tensor
```
```suggestion
  have hfromTensorFinsuppDegree := fromTensorFinsupp_degree (q:=q) tensor
```

The fact that we need to supply the `q` explicitly is probably telling us to make it an explicit parameter. @goens I leave it to you as to whether you want to do this in the same PR. I'd be happy with a follow-up PR that makes the `q` explicit here!

bollu reviewed (approved):
LGTM, thanks for the refactor :) I feel we should make `q` explicit in places where we pass it implicitly because it's clear Lean's not able to infer it; I'm happy for that to be in a different PR, or if you want to pass the ball to me to do this refactor!

goens committed:
make fromTensorFinsupp_degree parameter q explicit

goens committed:
forgotten a few qs

goens commented:
@bollu good point! I removed the cases for `fromTensorFinsupp_degree` which I had added in this PR to keep the diff small, we might consider a second pass where we do this for others like `R.fromTensor` itself. The thing that gives me pause there is that sometimes it does work, like on theorem `fromTensor_toTensor`, where it's cleaner with the implicits. I think in any case that one should be a separate PR
